nifi-nar-bundles/nifi-jms-bundle/nifi-jms-processors/src/main/java/org/apache/nifi/jms/cf/JndiJmsConnectionFactoryProvider.java

```diff
@@ -0,0 +1,165 @@
+        .displayName("Naming Provider URL")
+        .description("The URL of the JNDI Naming Provider to use")
+        .required(true)
+        .addValidator(NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .build();
+    static final PropertyDescriptor CONNECTION_FACTORY_NAME = new Builder()
+        .name("connection.factory.name")
+        .displayName("Connection Factory Name")
+        .description("The name of the JNDI Object to lookup for the Connection Factory")
+        .required(true)
+        .addValidator(NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .build();
+    static final PropertyDescriptor NAMING_FACTORY_LIBRARIES = new Builder()
+        .name("naming.factory.libraries")
+        .displayName("Naming Factory Libraries")
+        .description("Specifies .jar files or directories to add to the ClassPath in order to find the Initial Naming Factory Class")
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .addValidator(StandardValidators.createListValidator(true, true, StandardValidators.createURLorFileValidator()))
+        .dynamicallyModifiesClasspath(true)
+        .build();
+
+    private static final List<PropertyDescriptor> PROPERTY_DESCRIPTORS = Arrays.asList(
+        INITIAL_NAMING_FACTORY_CLASS,
+        NAMING_PROVIDER_URL,
+        CONNECTION_FACTORY_NAME,
+        NAMING_FACTORY_LIBRARIES);
+
+    private ConnectionFactory connectionFactory;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTY_DESCRIPTORS;
+    }
+
+    @Override
+    protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String propertyDescriptorName) {
+        return new Builder()
+            .name(propertyDescriptorName)
+            .displayName(propertyDescriptorName)
+            .description("JNDI Initial Context Environment configuration for '" + propertyDescriptorName + "'")
+            .required(false)
+            .dynamic(true)
+            .addValidator(Validator.VALID)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .build();
+    }
+
+    @OnDisabled
+    public void shutdown() {
+        connectionFactory = null;
+    }
+
+    @Override
+    public synchronized ConnectionFactory getConnectionFactory() {
+        if (connectionFactory == null) {
+            connectionFactory = lookupConnectionFactory();
+        } else {
+            getLogger().debug("Connection Factory has already been obtained from JNDI. Will return cached instance.");
+        }
+
+        return connectionFactory;
+    }
+
+    private ConnectionFactory lookupConnectionFactory() {
+        try {
+            final ConfigurationContext context = getConfigurationContext();
+
+            final Hashtable env = new Hashtable();
```

Comment:
Hashtable support generics. Can it be parametrized? 
final Hashtable<String,String> env = new Hashtable<>();

Reply:
Good call.